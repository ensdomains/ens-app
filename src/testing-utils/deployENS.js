module.exports = async function deployENS({ web3, accounts }) {
  const { sha3 } = web3.utils
  function deploy(contractJSON, ...args) {
    const contract = new web3.eth.Contract(contractJSON.abi)
    return contract
      .deploy({
        data: contractJSON.bytecode,
        arguments: args
      })
      .send({
        from: accounts[0],
        gas: 5700000
      })
  }

  function namehash(name) {
    let node =
      '0x0000000000000000000000000000000000000000000000000000000000000000'
    if (name !== '') {
      let labels = name.split('.')
      for (let i = labels.length - 1; i >= 0; i--) {
        node = sha3(node + sha3(labels[i]).slice(2), {
          encoding: 'hex'
        })
      }
    }
    return node.toString()
  }

  function loadContract(modName, contractName){
    return require(`@ensdomains/${modName}/build/contracts/${contractName}`)
  }
  
  const registryJSON = loadContract('ens', 'ENSRegistry')
  const resolverJSON = loadContract('resolver', 'PublicResolver')
  const oldResolverJSON = loadContract('ens-022', 'PublicResolver')
  const reverseRegistrarJSON = loadContract('ens', 'ReverseRegistrar')
  const testRegistrarJSON = loadContract('ens', 'TestRegistrar')
  const hashRegistrarSimplifiedJSON = loadContract('ens', 'HashRegistrar')

  /* Deploy the main contracts  */
  const ens = await deploy(registryJSON)
  const resolver = await deploy(resolverJSON, ens._address)
  const oldResolver = await deploy(oldResolverJSON, ens._address)
  const reverseRegistrar = await deploy(
    reverseRegistrarJSON,
    ens._address,
    resolver._address
  )

  const testRegistrar = await deploy(
    testRegistrarJSON,
    ens._address,
    namehash('test')
  )
  // Disabled for now as the deploy was throwing error and this is not in use.
  const ethRegistrar = await deploy(
    hashRegistrarSimplifiedJSON,
    ens._address,
    namehash('eth'),
    0,
  )

  const ensContract = ens.methods
  const resolverContract = resolver.methods
  const oldResolverContract = oldResolver.methods
  const reverseRegistrarContract = reverseRegistrar.methods
  const testRegistrarContract = testRegistrar.methods
  const ethRegistrarContract = ethRegistrar.methods

  console.log('ENS registry deployed at: ', ens._address)
  console.log('Public resolver deployed at: ', resolver._address)
  console.log('Old Public resolver deployed at: ', oldResolver._address)
  console.log('Reverse Registrar deployed at: ', reverseRegistrar._address)
  console.log('Test Registrar deployed at: ', testRegistrar._address)
  console.log('Auction Registrar deployed at: ', ethRegistrar._address)

  const tld = 'eth'
  const tldHash = sha3(tld)

  /* Setup the root TLD */
  await ensContract
    .setSubnodeOwner('0x00000000000000000000000000000000', tldHash, accounts[0])
    .send({
      from: accounts[0]
    })

  await ensContract
    .setSubnodeOwner('0x00000000000000000000000000000000', sha3('test'), accounts[0])
    .send({
      from: accounts[0]
    })

  await ensContract.setResolver(namehash(''), resolver._address).send({from: accounts[0]})
  await ensContract.setResolver(namehash('eth'), resolver._address).send({from: accounts[0]})
  await ensContract.setResolver(namehash('test'), resolver._address).send({from: accounts[0]})

  await ensContract
    .setSubnodeOwner('0x00000000000000000000000000000000', sha3('test'), testRegistrar._address)
    .send({
      from: accounts[0]
    })

  let rootOwner = await ensContract.owner('0x00000000000000000000000000000000').call() 

  console.log('testRegistrarContract.register', namehash('example'), testRegistrar._address); 
  let tx = await testRegistrarContract.register(sha3('example'), accounts[0]).send({from:accounts[0]})
  let exampleTestOwner = await ensContract.owner(namehash('example.test')).call()
  /* Setup the root reverse node */
  await ensContract
    .setSubnodeOwner(
      '0x00000000000000000000000000000000',
      sha3('reverse'),
      accounts[0]
    )
    .send({
      from: accounts[0]
    })

    await ensContract
    .setSubnodeOwner(
      namehash('reverse'),
      sha3('addr'),
      accounts[0]
    )
    .send({
      from: accounts[0]
    })

  await ensContract.setResolver(namehash('addr.reverse'), resolver._address).send({from: accounts[0]})
  


  /* Setup the reverse subdomain: addr.reverse */
  await ensContract
    .setSubnodeOwner(
      namehash('reverse'),
      sha3('addr'),
      reverseRegistrar._address
    )
    .send({
      from: accounts[0]
    })

  /* Register the subdomain resolver.eth */

  await ensContract
    .setSubnodeOwner(namehash('eth'), sha3('resolver'), accounts[0])
    .send({
      from: accounts[0]
    })

  await ensContract
    .setSubnodeOwner(namehash('eth'), sha3('oldresolver'), accounts[0])
    .send({
      from: accounts[0]
    })

  /* Register some test domains */

  await ensContract
    .setSubnodeOwner(namehash('eth'), sha3('awesome'), accounts[0])
    .send({
      from: accounts[0]
    })

  await ensContract
    .setSubnodeOwner(namehash('eth'), sha3('superawesome'), accounts[0])
    .send({
      from: accounts[0]
    })

  // Setup domain  with a resolver
  await ensContract
    .setSubnodeOwner(namehash('eth'), sha3('notsoawesome'), accounts[0])
    .send({
      from: accounts[0]
    })

  await ensContract
    .setResolver(namehash('notsoawesome.eth'), resolver._address)
    .send({
      from: accounts[0]
    })

  /* Setup domain with a resolver and addr/content */
  await ensContract
    .setSubnodeOwner(namehash('eth'), sha3('abittooawesome'), accounts[0])
    .send({
      from: accounts[0]
    })

  const aBitTooAwesome = namehash('abittooawesome.eth')

  await ensContract.setResolver(aBitTooAwesome, resolver._address).send({
    from: accounts[0]
  })
  await resolverContract.setAddr(aBitTooAwesome, resolver._address).send({
    from: accounts[0]
  })

  await reverseRegistrarContract
    .setName('abittooawesome.eth')
    .send({ from: accounts[0], gas: 1000000 })

  await resolverContract
    .setContenthash(
      aBitTooAwesome,
      '0x736f6d65436f6e74656e74000000000000000000000000000000000000000001'
    )
    .send({
      from: accounts[0]
    })

  /* Setup some domains for subdomain testing */
  await ensContract
    .setSubnodeOwner(namehash('eth'), sha3('subdomaindummy'), accounts[0])
    .send({
      from: accounts[0]
    })

  await ensContract
    .setSubnodeOwner(
      namehash('subdomaindummy.eth'),
      sha3('original'),
      accounts[0]
    )
    .send({
      from: accounts[0]
    })

  await ensContract
    .setSubnodeOwner(namehash('eth'), sha3('subdomain'), accounts[0])
    .send({
      from: accounts[0]
    })

  /* Point the resolver.eth's resolver to the public resolver */
  await ensContract
    .setResolver(namehash('resolver.eth'), resolver._address)
    .send({
      from: accounts[0]
    })
  await ensContract
    .setResolver(namehash('oldresolver.eth'), oldResolver._address)
    .send({
      from: accounts[0]
    })

  /* Resolve the resolver.eth address to the address of the public resolver */
  await resolverContract
    .setAddr(namehash('resolver.eth'), resolver._address)
    .send({
      from: accounts[0]
    })
  /* Resolve the oldresolver.eth address to the address of the public resolver */

  await resolverContract
    .setAddr(namehash('oldresolver.eth'), oldResolver._address)
    .send({
      from: accounts[0]
    })

  /* Resolve the resolver.eth content to a 32 byte content hash */

  await resolverContract
    .setContenthash(
      namehash('resolver.eth'),
      // ipfs://QmTeW79w7QQ6Npa3b1d5tANreCDxF2iDaAPsDvW6KtLmfB
      '0xe301017012204edd2984eeaf3ddf50bac238ec95c5713fb40b5e428b508fdbe55d3b9f155ffe'
    )
    .send({
      from: accounts[0], gas:5000000
    })

  await oldResolverContract
    .setContent(
      namehash('oldresolver.eth'),
      '0x736f6d65436f6e74656e74000000000000000000000000000000000000000000'
    )
    .send({
      from: accounts[0]
    })

    /* Setup a reverse for account[0] to eth tld  */

  await reverseRegistrarContract
    .setName('eth')
    .send({ from: accounts[2], gas: 1000000 })

  /* Set the registrar contract as the owner of .eth */
  // await ensContract
  //   .setOwner(namehash('eth'), accounts[0])
  //   .send({ from: accounts[0] })

  return {
    emptyAddress:'0x0000000000000000000000000000000000000000',
    ensAddress: ens._address,
    ownerAddress: accounts[0],
    resolverAddress: resolver._address,
    reverseRegistrarAddress: reverseRegistrar._address,
    reverseRegistrarOwnerAddress: accounts[0]
  }
}
