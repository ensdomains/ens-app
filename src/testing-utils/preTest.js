import { deployENS as deployTestEns } from '@ensdomains/mock'
import Web3 from 'web3'
import fs from 'fs'

let web3

export async function getAccounts(web3) {
  return web3.eth.getAccounts()
}

async function setupWeb3(customProvider) {
  web3 = new Web3(customProvider)
  const networkId = await web3.eth.net.getId()

  return {
    web3,
    networkId
  }
}

async function init() {
  const ENV = process.argv[2]

  switch (ENV) {
    case 'GANACHE_GUI':
      var provider = new Web3.providers.HttpProvider('http://localhost:7545')
      var { web3 } = await setupWeb3(provider)
      break
    case 'GANACHE_CLI':
      var provider = new Web3.providers.HttpProvider('http://localhost:8545')
      var { web3 } = await setupWeb3(provider)
      break
    default:
      var provider = new Web3.providers.HttpProvider('http://localhost:8545')
      var { web3 } = await setupWeb3(provider)
      break
  }

  const accounts = await getAccounts(web3)

  const addresses = await deployTestEns({ web3, accounts })
  const {
    ensAddress,
    controllerAddress,
    legacyAuctionRegistrarAddress
  } = addresses

  fs.writeFileSync('./cypress.env.json', JSON.stringify(addresses))
  fs.writeFile('./.env.local', `REACT_APP_ENS_ADDRESS=${ensAddress}`, err => {
    if (err) throw err
    console.log(`Successfully wrote ENS address ${ensAddress} to .env.local`)
  })
}

init()
