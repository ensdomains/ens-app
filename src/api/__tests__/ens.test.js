/**
 * @jest-environment node
 */
import GanacheCLI from 'ganache-cli'
import setupWeb3, { getAccounts } from '../web3'
import deployENS from '../../testing-utils/deployENS'
import {
  getOwner,
  setOwner,
  setSubnodeOwner,
  getResolver,
  setResolver,
  getAddr,
  setAddress,
  getContent,
  setContent,
  createSubdomain,
  deleteSubdomain,
  getDomainDetails,
  getSubDomains,
  getName,
  claimAndSetReverseRecordName
} from '../registry'
import getENS, { getNamehash } from '../ens'
import '../../testing-utils/extendExpect'
import Web3 from 'web3'

const ENVIRONMENTS = ['GANACHE_GUI', 'GANACHE_CLI', 'GANACHE_CLI_MANUAL']
const ENV = ENVIRONMENTS[1]

let reverseRegistrar

describe('Blockchain tests', () => {
  beforeAll(async () => {
    switch (ENV) {
      case 'GANACHE_CLI':
        var provider = GanacheCLI.provider()
        var web3 = await setupWeb3(provider)
        break
      case 'GANACHE_GUI':
        var provider = new Web3.providers.HttpProvider('http://localhost:7545')
        var web3 = await setupWeb3(provider)
        break
      case 'GANACHE_CLI_MANUAL':
        var provider = new Web3.providers.HttpProvider('http://localhost:8545')
        var web3 = await setupWeb3(provider)
        break
      default:
        const options = ENVIRONMENTS.join(' or ')
        throw new Error(`ENV not set properly, please pick from ${options}`)
    }

    const accounts = await getAccounts()
    expect(accounts.length).toBeGreaterThan(0)

    const { ensAddress, reverseRegistrarAddress } = await deployENS({
      web3,
      accounts
    })

    reverseRegistrar = reverseRegistrarAddress

    await getENS(ensAddress)
  }, 30000)

  describe('Test contract and Web3 setup', () => {
    test('accounts exist', async () => {
      const accounts = await getAccounts()
      expect(accounts.length).toBeGreaterThan(0)
    })

    test('ens registry, resolver and reverse registrar deployed', async () => {
      const { ENS } = await getENS()
      const accounts = await getAccounts()

      const eth = getNamehash('eth')
      const ethOwner = await ENS.owner(eth).call()
      expect(ethOwner).toBe(accounts[0])

      const reverse = getNamehash('reverse')
      const reverseOwner = await ENS.owner(reverse).call()
      expect(reverseOwner).toBe(accounts[0])

      const reverseNode = getNamehash('addr.reverse')
      const reverseNodeOwner = await ENS.owner(reverseNode).call()
      expect(reverseNodeOwner).toBe(reverseRegistrar)
    })
  })

  describe('Registry', () => {
    test('getOwner returns owner', async () => {
      const accounts = await getAccounts()
      const owner = await getOwner('eth')
      expect(owner).toBe(accounts[0])
    })

    test('setSubnodeOwner sets new subnode owner', async () => {
      const owner = await getOwner('subnode.resolver.eth')
      const accounts = await getAccounts()
      expect(owner).toBe('0x0000000000000000000000000000000000000000')
      await setSubnodeOwner('subnode', 'resolver.eth', accounts[0])
      const newOwner = await getOwner('subnode.resolver.eth')
      expect(newOwner).toBe(accounts[0])
    })

    test('setNewOwner sets new owner', async () => {
      const owner = await getOwner('givethisaway.awesome.eth')
      const accounts = await getAccounts()
      expect(owner).toBe('0x0000000000000000000000000000000000000000')
      await setSubnodeOwner('givethisaway', 'awesome.eth', accounts[0])
      const owner2 = await getOwner('givethisaway.awesome.eth')
      expect(owner2).toBe(accounts[0])
      await setOwner('givethisaway.awesome.eth', accounts[1])
      const newOwner = await getOwner('givethisaway.awesome.eth')
      expect(newOwner).toBe(accounts[1])
    })

    test('getResolver returns a resolver address when set', async () => {
      const resolver = await getResolver('resolver.eth')
      expect(resolver).toBeHex()
      expect(resolver).toBeEthAddress()
      expect(resolver).not.toBe('0x0000000000000000000000000000000000000000')
    })

    test('getResolver returns 0x00... when resolver address is not set', async () => {
      const resolver = await getResolver('eth')
      expect(resolver).toBeHex()
      expect(resolver).toBeEthAddress()
      expect(resolver).toBe('0x0000000000000000000000000000000000000000')
    })

    test('setResolver sets the resolver on a node', async () => {
      //test setResolver
      const resolver = await getResolver('awesome.eth')
      const mockResolver = '0x0000000000000000000000000000000000abcdef'
      expect(resolver).not.toBe(mockResolver)

      await setResolver('awesome.eth', mockResolver)
      const newResolver = await getResolver('awesome.eth')
      expect(newResolver).toBeHex()
      expect(newResolver).toBeEthAddress()
      expect(newResolver.toLowerCase()).toBe(mockResolver)
    })

    test('createSubdomain makes a new subdomain', async () => {
      const accounts = await getAccounts()
      const oldOwner = await getOwner('a.subdomain.eth')
      // expect the initial owner to be no one
      expect(oldOwner).toBe('0x0000000000000000000000000000000000000000')
      await createSubdomain('subdomain', 'eth')
      const newOwner = await getOwner('subdomain.eth')
      // Verify owner is the user and therefore the subdomain exists
      expect(newOwner).toBe(accounts[0])
    })

    test('deleteSubdomain deletes a subdomain', async () => {
      const accounts = await getAccounts()
      const oldOwner = await getOwner('b.subdomain.eth')
      // expect the initial owner to be no one
      expect(oldOwner).toBe('0x0000000000000000000000000000000000000000')
      await createSubdomain('b', 'subdomain.eth')
      const newOwner = await getOwner('b.subdomain.eth')
      // Verify owner is the user and therefore the subdomain exists
      expect(newOwner).toBe(accounts[0])
      await deleteSubdomain('b', 'subdomain.eth')
      const deletedOwner = await getOwner('b.subdomain.eth')
      // Verify owner has been set to 0x00... to ensure deletion
      expect(deletedOwner).toBe('0x0000000000000000000000000000000000000000')
    })
  })

  describe('Resolver', () => {
    test('getAddr returns an address', async () => {
      const addr = await getAddr('resolver.eth')
      expect(addr).toBeHex()
      expect(addr).toBeEthAddress()
      expect(addr).not.toBe('0x0000000000000000000000000000000000000000')
    })

    test('getAddr returns 0x000', async () => {
      await createSubdomain('bar', 'eth')
      const resolverAddr = await getAddr('resolver.eth')
      await setResolver('bar.eth', resolverAddr)
      const addr = await getAddr('bar.eth')
      expect(addr).toBe('0x0000000000000000000000000000000000000000')
    })

    test('setAddr sets an address', async () => {
      //reverts if no addr is present
      const resolverAddr = await getAddr('resolver.eth')
      await setResolver('superawesome.eth', resolverAddr)
      await setAddress(
        'superawesome.eth',
        '0x0000000000000000000000000000000000012345'
      )
      const addr = await getAddr('superawesome.eth')
      expect(addr).toBe('0x0000000000000000000000000000000000012345')
    })

    test('getContent returns a 32 byte hash', async () => {
      const content = await getContent('resolver.eth')
      expect(content).toBeHex()
      expect(content).toMatchSnapshot()
    })

    test('getContent returns 0x00... when no content has been set', async () => {
      const content = await getContent('superawesome.eth')
      expect(content).toBe(
        '0x0000000000000000000000000000000000000000000000000000000000000000'
      )
    })

    test('setContent sets 32 byte hash', async () => {
      await setContent(
        'superawesome.eth',
        '0x736f6d65436f6e74656e74000000000000000000000000000000000000000000'
      )

      const content = await getContent('superawesome.eth')
      expect(content).toBeHex()
      expect(content).toMatchSnapshot()
    })
  })

  describe('Reverse Registrar', () => {
    test('reverseNode is owned by reverseRegistrar', async () => {
      const owner = await getOwner('addr.reverse')
      expect(reverseRegistrar).toBe(owner)
    })

    test('getName gets a name for an address', async () => {
      const accounts = await getAccounts()
      const { name } = await getName(accounts[2])
      expect(name).toBe('eth')
    })

    test('claimAndSetReverseRecordName claims and sets a name', async () => {
      const accounts = await getAccounts()
      const { name } = await getName(accounts[0])
      expect(name).toBe(null)
      await claimAndSetReverseRecordName('eth', 200000)
      const { name: nameAfter } = await getName(accounts[0])
      expect(nameAfter).toBe('eth')
    })
  })

  describe('Helper functions', () => {
    test('getDomainDetails gets rootdomain and resolver details', async () => {
      const domain = await getDomainDetails('resolver.eth')
      const accounts = await getAccounts()
      expect(domain.owner).not.toBe(
        '0x0000000000000000000000000000000000000000'
      )
      expect(domain.owner).toBeEthAddress()
      expect(domain.resolver).not.toBe(
        '0x0000000000000000000000000000000000000000'
      )
      expect(domain.resolver).toBeEthAddress()
      const addr = await getAddr('resolver.eth')
      expect(domain.addr).toBe(addr)
      expect(domain.content).toMatchSnapshot()
    })

    test('getSubdomains gets all subdomains', async () => {
      const domains = await getSubDomains('eth')
      expect(domains.length).toBeGreaterThan(0)
      expect(domains[domains.length - 1].label).toBe('resolver')
    })
  })
})
