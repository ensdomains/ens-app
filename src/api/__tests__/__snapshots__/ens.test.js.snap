// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blockchain tests Helper functions getDomainDetails gets rootdomain and resolver details 1`] = `"0x3132333435363738390000000000000000000000000000000000000000000000"`;

exports[`Blockchain tests Helper functions getRootDomain gets rootdomain and resolver details 1`] = `"0x3132333435363738390000000000000000000000000000000000000000000000"`;

exports[`Blockchain tests Resolver getContent returns a 32 byte hash 1`] = `"0x3132333435363738390000000000000000000000000000000000000000000000"`;

exports[`Blockchain tests Resolver setContent sets 32 byte hash 1`] = `"0x736f6d65436f6e74656e74000000000000000000000000000000000000000000"`;
